core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,255 @@
+        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, factory);
+    }
+
+    public static void input(Type type, ListaggAggregationState state, Block value, Slice separator, boolean overflowError, Slice overflowFiller, boolean showOverflowEntryCount, int position)
+    {
+        if (state.isEmpty()) {
+            if (overflowFiller.length() > MAX_OVERFLOW_FILLER_LENGTH) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("Overflow filler length %d exceeds maximum length %d", overflowFiller.length(), MAX_OVERFLOW_FILLER_LENGTH));
+            }
+            // Set the parameters of the LISTAGG command within the state so that
+            // they can be used within the `output` function
+            state.setSeparator(separator);
+            state.setOverflowError(overflowError);
+            state.setOverflowFiller(overflowFiller);
+            state.setShowOverflowEntryCount(showOverflowEntryCount);
+        }
+        if (!value.isNull(position)) {
+            state.add(value, position);
+        }
+    }
+
+    public static void combine(Type type, ListaggAggregationState state, ListaggAggregationState otherState)
+    {
+        Slice previousSeparator = state.getSeparator();
+        if (previousSeparator == null) {
+            state.setSeparator(otherState.getSeparator());
+            state.setOverflowError(otherState.isOverflowError());
+            state.setOverflowFiller(otherState.getOverflowFiller());
+            state.setShowOverflowEntryCount(otherState.showOverflowEntryCount());
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        int separatorLength = separator.length();
+        OutputContext outputContext = new OutputContext();
+        state.forEach((block, position) -> {
+            if (outputContext.outputLength > maxOutputLength) {
+                outputContext.outputLength += separatorLength + block.getSliceLength(position);
+            }
+            else {
+                if (outputContext.emittedEntryCount > 0) {
+                    if (outputContext.outputLength + separatorLength <= maxOutputLength) {
+                        // write the separator only if it fully fits within maxOutputLength
+                        out.writeBytes(separator, 0, separatorLength);
+                        outputContext.outputLength += separatorLength;
+                    }
+                    else {
+                        outputContext.outputLength += separatorLength;
```

Comment:
This seems incorrect, since the separator is not added in this branch.

Reply:
I notice that `outputLength` is actually misleading in this context.

I intended to store in `outputLength` the value shown eventually to the user in case of an overflow error.
I'll refactor slightly the implementation to avoid leaving place for doubts by adding a boolean `overflow` flag to the output context.